### DISCUSSION TITLE
Garbled command substitution in nyagos' built-in ls

### DISCUSSION THREAD

**Discussion by kgasawa** [nyaosorg/nyagos#433] [Q&A]
GitHubで初めて質問させていただきます。作法違いがございまし
たらすみません。

　nyagosの内蔵lsの出力をコマンド出力置換した際に、文字化けが
生じました。これについて質問させてください。 

　具体的には、なかに全角文字を含むファイル名が置かれているディ
レクトリ内で、内蔵lsの出力をコマンド出力置換したときに文字化け
が生じます。

例）
```
　echo `ls`
```
　あるいは
```
　echo $(ls)
```

　これは内蔵lsコマンドの仕様で文字化けを避けられないのでしょ
うか？あるいは何か設定することなどにより回避できますでしょ
うか？

　なお実際はforeach文でコマンド出力置換を用いてlsの結果を展開
し、続く文でコマンド処理をしようとしていました。一応、lsの代
わりにdir /Bを用いることで意図する作業はできたものの、lsの仕様
上はどのようになっているのかと思い質問させていただきました。

　以下補足事項です。
●手元の環境
　Windows 11上で、nyagosのamd版（4.4.13_3）を使用しております。

　どうぞよろしくお願いいたします。

**Comment by hymkor** [comment]
ご報告ありがとうございます。その症状は

- 内蔵 ls は UTF8 で出力する
- `` `～` `` と `$(～)` は出力を「非UTF8」(日本の場合SJIS)と想定して、UTF8への変換をしている

と齟齬が発生しているために発生しているようです。
（すでにUTF8になっているテキストをさらにUTF8変換しようとしているので化ける）

個々のレベルでは仕様動作ではあるのですが、組み合わせてうまく動作しないというのもどうかと思われますので、

- `` `～` `` と `$(～)` の出力が明らかに UTF8 でない場合のみ、UTF8 変換する

という形でうまくできないか、実験してみます。
（理屈の上では 100% うまくゆくわけではないが、9割方いけるはず）

**Reply by kgasawa** [reply]
ご回答いただきありがとうございます。文字化け発生の背景がよくわかりました。
また何かわかりましたら、よろしくお願いいたします。

**Comment by hymkor** [comment]
* `` `～` `` と `$(～)` の出力が明らかに UTF8 でない場合のみ、UTF8 変換する

で対応しました。

- d289e8f1b561dd9b660b5844368c0836fbd40906 引数が不適切な UTF8 の時は ANSI(SJIS)→UTF8 変換した文字列を戻り値として返し、そうでない時はそのまま返すLua関数 `nyagos.atou_if_needed` を実装
- 6da0a65b540389e31d543b70e1b5eb8a202da752 逆クォート（nyagos.d/backquote.lua）で `atou` （ANSI(SJIS)→UTF8変換関数） ではなく、`atou_if_needed` を使う

一応、これで100%ではありませんが、9割9分の文字化けは回避できる見込みです。
（`%USERPROFILE%\Documents` 下での、内蔵ls と外部のls.exe の出力を引用して確認）

次のリリース（2023年下期版：9～10月頃）でパッケージに反映する予定です。

**Reply by kgasawa** [reply]
ご対応いただき、ありがとうございます。次のリリースをお待ち申し上げております。取り急ぎ、御礼まで。

**Comment by hymkor** [comment]
たいへん、お待たせいたしました。[4.4.14_0](https://github.com/nyaosorg/nyagos/releases/tag/4.4.14_0) にて修正をバイナリに反映いたしました。本件に関するところは次の２点になります。

> - 適切なUTF8文字列でない時は ANSI文字列とみなして UTF8変換を試みる関数 nyagos.atou_if_needed を追加
> - [#433](https://github.com/nyaosorg/nyagos/discussions/433) 文字化けを避けるために、逆クォートでは nyagos.atou_if_needed を使って、UTF8 を更に UTF8 化させないようにした

以上、よろしくお願いいたします。

**Comment by hymkor** [comment]
本件、クローズさせていただきます。
何か問題などありましたら、再オープンください。

**Comment by kgasawa** [comment]
葉山薫 様：

　お世話になっております、我澤です。
　最新版公開後の確認、報告、御礼が遅くなり、申し訳ござい
ませんでした。

　nyagos 4.4.14_0 をダウンロード・インストールし、問題が
解決していることを確認いたしました。あらためて、ありがと
うございました。

我澤賢之

**************************************************
我澤 賢之（がさわ・けんじ）
国立障害者リハビリテーションセンター
研究所 障害福祉研究部 主任研究官
企画・情報部 支援機器イノベーション情報・支援室
tel 04-2995-3100(内線2515/(PHS)7179)
fax 04-2995-3132
***@***.***
**************************************************
